### PR TITLE
Move accounting, reporting, and statistical methods out of Campaign

### PR DESCRIPTION
**Pull request by sixlettervariables**
This removes accounting, reporting, and statistical methods out of Campaign and either into (a) a new class more closely following SRP, or (b) the only class containing the only callers of the method.

### CODE REVIEW

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+                continue;
+            }
+            if (unit.isMothballed()) {
+                if (type == Unit.ETYPE_MOTHBALLED) {
+                    num++;
+                }
+                continue;
+            }
+            Entity en = unit.getEntity();
+            if (en instanceof GunEmplacement || en instanceof FighterSquadron || en instanceof Jumpship) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:62]:
```suggestion
            if ((en instanceof GunEmplacement) || (en instanceof FighterSquadron) || (en instanceof Jumpship)) {
```

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+                continue;
+            }
+            if (type == Entity.ETYPE_MECH && en instanceof Mech) {
+                num++;
+                continue;
+            }
+            if (type == Entity.ETYPE_DROPSHIP && en instanceof Dropship) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:69]:
```suggestion
            if ((type == Entity.ETYPE_DROPSHIP) && (en instanceof Dropship)) {
```

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+            if (type == Entity.ETYPE_CONV_FIGHTER && en instanceof ConvFighter) {
+                num++;
+                continue;
+            }
+            if (type == Entity.ETYPE_AERO && en instanceof Aero
+                    && !(en instanceof SmallCraft || en instanceof ConvFighter)) {
+                num++;
+                continue;
+            }
+            if (type == Entity.ETYPE_INFANTRY && en instanceof Infantry && !(en instanceof BattleArmor)) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:87]:
```suggestion
            if ((type == Entity.ETYPE_INFANTRY) && (en instanceof Infantry) && !(en instanceof BattleArmor)) {
```

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+                num++;
+                continue;
+            }
+            if (type == Entity.ETYPE_BATTLEARMOR && en instanceof BattleArmor) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:91]:
```suggestion
            if ((type == Entity.ETYPE_BATTLEARMOR) && (en instanceof BattleArmor)) {
```

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+    public int getOccupiedBays(long type, boolean lv) {
+        int num = 0;
+        for (Unit unit : getHangar().getUnits()) {
+            if (unit.isMothballed()) {
+                continue;
+            }
+            Entity en = unit.getEntity();
+            if (en instanceof GunEmplacement || en instanceof Jumpship) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:120]:
```suggestion
            if ((en instanceof GunEmplacement) || (en instanceof Jumpship)) {
```

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+                continue;
+            }
+            if (type == Entity.ETYPE_MECH && en instanceof Mech) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:123]:
```suggestion
            if ((type == Entity.ETYPE_MECH) && (en instanceof Mech)) {
```

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+                num++;
+                continue;
+            }
+            if (type == Entity.ETYPE_DROPSHIP && en instanceof Dropship) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:127]:
```suggestion
            if ((type == Entity.ETYPE_DROPSHIP) && (en instanceof Dropship)) {
```

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+                num++;
+                continue;
+            }
+            if (type == Entity.ETYPE_SMALL_CRAFT && en instanceof SmallCraft && !(en instanceof Dropship)) {
+                num++;
+                continue;
+            }
+            if (type == Entity.ETYPE_CONV_FIGHTER && en instanceof ConvFighter) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:135]:
```suggestion
            if ((type == Entity.ETYPE_CONV_FIGHTER) && (en instanceof ConvFighter)) {
```

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+                num++;
+                continue;
+            }
+            if (type == Entity.ETYPE_AERO && en instanceof Aero
+                    && !(en instanceof SmallCraft || en instanceof ConvFighter)) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:140]:
```suggestion
            if ((type == Entity.ETYPE_AERO) && (en instanceof Aero)
                    && !((en instanceof SmallCraft) || (en instanceof ConvFighter))) {
```

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+            }
+            if (type == Entity.ETYPE_INFANTRY && en instanceof Infantry && !(en instanceof BattleArmor)) {
+                num++;
+                continue;
+            }
+            if (type == Entity.ETYPE_BATTLEARMOR && en instanceof BattleArmor) {
+                num++;
+                continue;
+            }
+            if (type == Entity.ETYPE_TANK && en instanceof Tank) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:152]:
```suggestion
            if ((type == Entity.ETYPE_TANK) && (en instanceof Tank)) {
```

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+                if ((en.getWeight() <= 50 && lv) || (en.getWeight() > 50 && !lv)) {
+                    num++;
+                }
+                continue;
+            }
+            if (type == Entity.ETYPE_PROTOMECH && en instanceof Protomech) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:158]:
```suggestion
            if ((type == Entity.ETYPE_PROTOMECH) && (en instanceof Protomech)) {
```

File: MekHQ/src/mekhq/campaign/CampaignSummary.java [hunk 12]
```diff
@@ -146,41 +148,43 @@ public void updateInformation() {
...
         @SuppressWarnings("unused") // FIXME: What type of bays do ConvFighters use?
-                int noCF = Math.max(campaign.getNumberOfUnitsByType(Entity.ETYPE_CONV_FIGHTER) - campaign.getOccupiedBays(Entity.ETYPE_CONV_FIGHTER), 0);
-        int noASF = Math.max(campaign.getNumberOfUnitsByType(Entity.ETYPE_AERO) - campaign.getOccupiedBays(Entity.ETYPE_AERO), 0);
-        int nolv = Math.max(campaign.getNumberOfUnitsByType(Entity.ETYPE_TANK, false, true) - campaign.getOccupiedBays(Entity.ETYPE_TANK, true), 0);
-        int nohv = Math.max(campaign.getNumberOfUnitsByType(Entity.ETYPE_TANK) - campaign.getOccupiedBays(Entity.ETYPE_TANK), 0);
-        int noinf = Math.max(campaign.getNumberOfUnitsByType(Entity.ETYPE_INFANTRY) - campaign.getOccupiedBays(Entity.ETYPE_INFANTRY), 0);
-        int noBA = Math.max(campaign.getNumberOfUnitsByType(Entity.ETYPE_BATTLEARMOR) - campaign.getOccupiedBays(Entity.ETYPE_BATTLEARMOR), 0);
+                int noCF = Math.max(hangarStats.getNumberOfUnitsByType(Entity.ETYPE_CONV_FIGHTER) - hangarStats.getOccupiedBays(Entity.ETYPE_CONV_FIGHTER), 0);
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/CampaignSummary.java:162]: This looks weird

File: MekHQ/src/mekhq/campaign/report/PersonnelReport.java [hunk 13]
```diff
@@ -27,53 +27,184 @@
...
+        int dependents = 0;
+
+        for (Person p : getCampaign().getPersonnel()) {
+            // Add them to the total count
+            boolean primarySupport = p.hasPrimarySupportRole(false);
+
+            if (primarySupport && p.getPrisonerStatus().isFree() && p.getStatus().isActive()) {
+                countPersonByType[p.getPrimaryRole()]++;
+                countTotal++;
+                if (p.getInjuries().size() > 0 || p.getHits() > 0) {
```
Comment (Windchild292) [inline MekHQ/src/mekhq/campaign/report/PersonnelReport.java:156]:
```suggestion
                if ((p.getInjuries().size() > 0) || (p.getHits() > 0)) {
```

File: MekHQ/src/mekhq/campaign/CampaignSummary.java [hunk 14]
```diff
@@ -146,41 +148,43 @@ public void updateInformation() {
...
+        int nohv = Math.max(hangarStats.getNumberOfUnitsByType(Entity.ETYPE_TANK) - hangarStats.getOccupiedBays(Entity.ETYPE_TANK), 0);
+        int noinf = Math.max(hangarStats.getNumberOfUnitsByType(Entity.ETYPE_INFANTRY) - hangarStats.getOccupiedBays(Entity.ETYPE_INFANTRY), 0);
+        int noBA = Math.max(hangarStats.getNumberOfUnitsByType(Entity.ETYPE_BATTLEARMOR) - hangarStats.getOccupiedBays(Entity.ETYPE_BATTLEARMOR), 0);
         @SuppressWarnings("unused") // FIXME: This should be used somewhere...
-                int noProto = Math.max(campaign.getNumberOfUnitsByType(Entity.ETYPE_PROTOMECH) - campaign.getOccupiedBays(Entity.ETYPE_PROTOMECH),  0);
-        int freehv = Math.max(campaign.getTotalHeavyVehicleBays() - campaign.getOccupiedBays(Entity.ETYPE_TANK), 0);
-        int freeSC = Math.max(campaign.getTotalSmallCraftBays() - campaign.getOccupiedBays(Entity.ETYPE_SMALL_CRAFT), 0);
+                int noProto = Math.max(hangarStats.getNumberOfUnitsByType(Entity.ETYPE_PROTOMECH) - hangarStats.getOccupiedBays(Entity.ETYPE_PROTOMECH),  0);
```
[the remark this answers is not in the document]
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/CampaignSummary.java:169]: Probably is! I'll try and figure out why it's like this.

File: MekHQ/src/mekhq/campaign/unit/HangarStatistics.java [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+        for (Unit unit : getHangar().getUnits()) {
+            if (!inTransit && !unit.isPresent()) {
+                continue;
+            }
+            if (unit.isMothballed()) {
+                if (type == Unit.ETYPE_MOTHBALLED) {
+                    num++;
+                }
+                continue;
+            }
```
[the remark this answers is not in the document]
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/unit/HangarStatistics.java:60]: Probably is! I'll try and figure out why it's like this.

File: MekHQ/src/mekhq/campaign/CampaignSummary.java (hunk 12 above)
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/CampaignSummary.java:162]: Probably is! I'll try and figure out why it's like this.

File: MekHQ/src/mekhq/campaign/CampaignSummary.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (sixlettervariables) [reply MekHQ/src/mekhq/campaign/CampaignSummary.java:169]: This code has been there for a while, not sure exactly what its doing with the calcs but they were approved as valid back when this was introduced. I don't believe I changed the math any.

**Pull request merged**